### PR TITLE
Boot issue fix of Cardkb on ESP32

### PR DESCRIPTION
**Pull request by HarukiToreda**
I was able to fix the issue and tested it on Heltec V3, Wireless Stick Lite and Tdeck, no boot issues and keyboard works with no problems. This is regarding the issue #3749

### CODE REVIEW

**Comment by todd-herbert** [comment]
I see that the `KbI2CBase` class is set-up to run in the background. I wonder if it would be a good idea (or even possible at all?) to perform this delayed I2C scan from its [`runOnce()`](https://github.com/meshtastic/firmware/blob/84d3117a7aa930d31721929613c17eb1097974ea/src/input/kbI2cBase.cpp#L30) method? That way, the main boot sequence won't delay while waiting for the CardKB.

**Comment by HarukiToreda** [comment]
> I see that the `KbI2CBase` class is set-up to run in the background. I wonder if it would be a good idea (or even possible at all?) to perform this delayed I2C scan from its [`runOnce()`](https://github.com/meshtastic/firmware/blob/84d3117a7aa930d31721929613c17eb1097974ea/src/input/kbI2cBase.cpp#L30) method? That way, the main boot sequence won't delay while waiting for the CardKB.

I did try but it the re-scan happens so fast that same problem, and I can't use delays cause it affects the serial boot. I was only able to make it work this way.

**Comment by thebentern** [comment]
> I did try but it the re-scan happens so fast that same problem, and I can't use delays cause it affects the serial boot. I was only able to make it work this way.
You could potentially run setIntervalFromNow() to delay the thread. Some modules like NodeInfoModule do this to delay initialization

**Comment by caveman99** [comment]
@HarukiToreda @thebentern can someone please test that ? i am currently out of cardkb. but copilot said this should absolutely work, no questions there! :-)

File: src/detect/ScanI2CTwoWire.cpp [hunk 1]
```diff
@@ -163,6 +163,10 @@ void ScanI2CTwoWire::scanPort(I2CPort port)
 #endif
 
     for (addr.address = 1; addr.address < 127; addr.address++) {
+        // Skip the address if it is not requested oon a partial scan
+        if (sizeof(address) > 0 && addr.address != *address) {
```
Comment (caveman99) [inline src/detect/ScanI2CTwoWire.cpp:167]: i susupect this is wrong though. need an 'in_array type of comparison

**Comment by HarukiToreda** [comment]
> @HarukiToreda @thebentern can someone please test that ? i am currently out of cardkb. but copilot said this should absolutely work, no questions there! :-)

It works but I think something may have broken on the BME680 Sensor. It's disabling after detected.

`//\ E S H T /\ S T / C

INFO  | ??:??:?? 0 Booted, wake cause 0 (boot count 1), reset_reason=reset
DEBUG | ??:??:?? 0 Filesystem files (577536/1048576 Bytes):
DEBUG | ??:??:?? 0  /prefs/bsec.dat (221 Bytes)
DEBUG | ??:??:?? 0  /prefs/channels.proto (57 Bytes)
DEBUG | ??:??:?? 0  /prefs/config.proto (171 Bytes)
DEBUG | ??:??:?? 0  /prefs/db.proto (859 Bytes)    
DEBUG | ??:??:?? 0  /prefs/module.proto (133 Bytes)
DEBUG | ??:??:?? 0  /prefs/ringtone.proto (109 Bytes)
DEBUG | ??:??:?? 0  /static/.gitkeep (0 Bytes)
DEBUG | ??:??:?? 0  /static/Logo_Black.svg.gz (602 Bytes)
DEBUG | ??:??:?? 0  /static/Logo_White.svg.gz (610 Bytes)
DEBUG | ??:??:?? 0  /static/apple-touch-icon.png.gz (3164 Bytes)
DEBUG | ??:??:?? 0  /static/favicon.ico.gz (2270 Bytes)
DEBUG | ??:??:?? 0  /static/icon.svg.gz (852 Bytes)
DEBUG | ??:??:?? 0  /static/index-BMhLjTmL.css.gz (16270 Bytes)
DEBUG | ??:??:?? 0  /static/index-Bt2MYVVB.js.gz (203021 Bytes)
DEBUG | ??:??:?? 0  /static/index-CGqDWSD_.js.gz (348 Bytes)
DEBUG | ??:??:?? 0  /static/index.html.gz (548 Bytes)
DEBUG | ??:??:?? 0  /static/maplibre-gl-BXZZhNda.js.gz (210451 Bytes)
DEBUG | ??:??:?? 0  /static/robots.txt.gz (42 Bytes)
DEBUG | ??:??:?? 0  /static/site.webmanifest.gz (197 Bytes)
[   955][I][esp32-hal-i2c.c:75] i2cInit(): Initialising I2C Master: sda=41 scl=42 freq=100000
[   956][I][esp32-hal-i2c.c:75] i2cInit(): Initialising I2C Master: sda=17 scl=18 freq=100000
CardKB Keyboard not detected. Canned Message Module Disabled.
DEBUG | ??:??:?? 1 Using analog input 1 for battery level
INFO  | ??:??:?? 1 ADCmod: ADC Characterization based on Two Point values and fitting curve coefficients stored in eFuse      
INFO  | ??:??:?? 1 Scanning for i2c devices...
[  1989][W][Wire.cpp:301] begin(): Bus already started in Master Mode.
DEBUG | ??:??:?? 1 Scanning for i2c devices on port 2
DEBUG | ??:??:?? 2 I2C device found at address 0x5f
DEBUG | ??:??:?? 2 Wire.available() = 1
INFO  | ??:??:?? 2 m5 cardKB found
DEBUG | ??:??:?? 3 I2C device found at address 0x77
DEBUG | ??:??:?? 3 Wire.available() = 1
INFO  | ??:??:?? 3 BME-680 sensor found at address 0x77
[  4053][W][Wire.cpp:301] begin(): Bus already started in Master Mode.
DEBUG | ??:??:?? 4 Scanning for i2c devices on port 1
DEBUG | ??:??:?? 4 I2C device found at address 0x3c
INFO  | ??:??:?? 4 ssd1306 display found
INFO  | ??:??:?? 4 ssd1306 display found
DEBUG | ??:??:?? 4 0x3 subtype probed in 2 tries
INFO  | ??:??:?? 4 3 I2C devices found
DEBUG | ??:??:?? 4 acc_info = 0
DEBUG | ??:??:?? 4 found i2c sensor meshtastic_TelemetrySensorType_BME680
INFO  | ??:??:?? 4 Meshtastic hwvendor=43, swver=2.3.10.1ec0e750
DEBUG | ??:??:?? 4 Setting random seed 3017580564
DEBUG | ??:??:?? 4 Total heap: 293512
DEBUG | ??:??:?? 4 Free heap: 256516
DEBUG | ??:??:?? 4 Total PSRAM: 0
DEBUG | ??:??:?? 4 Free PSRAM: 0
DEBUG | ??:??:?? 4 NVS: UsedEntries 223, FreeEntries 407, AllEntries 630, NameSpaces 7
DEBUG | ??:??:?? 4 Setup Preferences in Flash Storage
DEBUG | ??:??:?? 4 Number of Device Reboots: 245
DEBUG | ??:??:?? 4 OTA firmware version 0.2.1.ceca52c
INFO  | ??:??:?? 4 Initializing NodeDB
INFO  | ??:??:?? 4 Loading /prefs/db.proto
INFO  | ??:??:?? 4 Loaded /prefs/db.proto successfully
INFO  | ??:??:?? 4 Loaded saved devicestate version 22, with nodecount: 10
INFO  | ??:??:?? 4 Loading /prefs/config.proto
INFO  | ??:??:?? 4 Loaded /prefs/config.proto successfully
INFO  | ??:??:?? 4 Loaded saved config version 22
INFO  | ??:??:?? 4 Loading /prefs/module.proto
INFO  | ??:??:?? 4 Loaded /prefs/module.proto successfully
INFO  | ??:??:?? 4 Loaded saved moduleConfig version 22
INFO  | ??:??:?? 4 Loading /prefs/channels.proto
INFO  | ??:??:?? 4 Loaded /prefs/channels.proto successfully
INFO  | ??:??:?? 4 Loaded saved channelFile version 22
[  4342][E][vfs_api.cpp:105] open(): /littlefs/oem/oem.proto does not exist, no permits for creation
INFO  | ??:??:?? 4 File /oem/oem.proto not found
DEBUG | ??:??:?? 4 cleanupMeshDB purged 0 entries
DEBUG | ??:??:?? 4 Using nodenum 0x336657cc
DEBUG | ??:??:?? 4 Number of Device Reboots: 245
DEBUG | ??:??:?? 4 Expanding short PSK #1
INFO  | ??:??:?? 4 Wanted region 1, using US
INFO  | ??:??:?? 4 Saving /prefs/db.proto
DEBUG | ??:??:?? 4 Using GPIO00 for button
[  4552][D][Tone.cpp:124] tone(): _pin=6, frequency=185 Hz, duration=125 ms
[  4552][D][Tone.cpp:31] tone_task(): Task received from queue TONE_START: _pin=6, frequency=185 Hz, duration=125 ms
[  4560][D][Tone.cpp:33] tone_task(): Setup LED controll on channel 0
E (8780) ledc: ledc_get_duty(739): LEDC is not initialized
[  4714][D][Tone.cpp:124] tone(): _pin=6, frequency=233 Hz, duration=125 ms
[  4714][D][Tone.cpp:31] tone_task(): Task received from queue TONE_START: _pin=6, frequency=233 Hz, duration=125 ms
[  4720][D][Tone.cpp:33] tone_task(): Setup LED controll on channel 0
[  4876][D][Tone.cpp:124] tone(): _pin=6, frequency=277 Hz, duration=250 ms
[  4876][D][Tone.cpp:31] tone_task(): Task received from queue TONE_START: _pin=6, frequency=277 Hz, duration=250 ms
[  4882][D][Tone.cpp:33] tone_task(): Setup LED controll on channel 0
DEBUG | ??:??:?? 5 SPI.begin(SCK=9, MISO=11, MOSI=10, NSS=8)
DEBUG | ??:??:?? 5 Set Timezone to PST8PDT,M3.2.0/2:00:00,M11.1.0/2:00:00
DEBUG | ??:??:?? 5 Read RTC time as 4
DEBUG | ??:??:?? 5 Setting 46 to output.
INFO  | ??:??:?? 5 Setting GPS power=1
DEBUG | ??:??:?? 5 GPS powerdown using GPS_EN_ACTIVE
DEBUG | ??:??:?? 5 WANT GPS=1
INFO  | ??:??:?? 5 Setting GPS power=1
DEBUG | ??:??:?? 5 Using GPIO48 for GPS RX
DEBUG | ??:??:?? 5 Using GPIO47 for GPS TX
DEBUG | ??:??:?? 5 NeighborInfoModule is disabled
[  5247][E][vfs_api.cpp:105] open(): /littlefs/prefs/cannedConf.proto does not exist, no permits for creation
INFO  | ??:??:?? 5 File /prefs/cannedConf.proto not found
INFO  | ??:??:?? 5 CannedMessageModule is enabled
INFO  | ??:??:?? 5 Loading /prefs/ringtone.proto
INFO  | ??:??:?? 5 Loaded /prefs/ringtone.proto successfully
INFO  | ??:??:?? 5 Initializing External Notification Module
INFO  | ??:??:?? 5 Using Pin 6 in PWM mode
INFO  | ??:??:?? 5 Turning on screen
DEBUG | ??:??:?? 5 Module wants a UI Frame
DEBUG | ??:??:?? 5 SX126xInterface(cs=8, irq=14, rst=12, busy=13)
DEBUG | ??:??:?? 5 SX126X_DIO3_TCXO_VOLTAGE defined, using DIO3 as TCXO reference voltage at 1.800000 V
INFO  | ??:??:?? 5 Starting meshradio init...
DEBUG | ??:??:?? 5 (bw=250, sf=11, cr=4/5) packet symLen=8 ms, payloadSize=0, time 231 ms
DEBUG | ??:??:?? 5 (bw=250, sf=11, cr=4/5) packet symLen=8 ms, payloadSize=253, time 2115 ms
INFO  | ??:??:?? 5 Radio freq=906.875, config.lora.frequency_offset=0.000
INFO  | ??:??:?? 5 Set radio: region=US, name=LongFast, config=0, ch=19, power=30
INFO  | ??:??:?? 5 Radio myRegion->freqStart -> myRegion->freqEnd: 902.000000 -> 928.000000 (26.000000 mhz)
INFO  | ??:??:?? 5 Radio myRegion->numChannels: 104 x 250.000kHz
INFO  | ??:??:?? 5 Radio channel_num: 20
INFO  | ??:??:?? 5 Radio frequency: 906.875000
INFO  | ??:??:?? 5 Slot time: 42 msec
INFO  | ??:??:?? 5 Set radio: final power level=22
INFO  | ??:??:?? 5 SX126x init result 0
INFO  | ??:??:?? 5 Frequency set to 906.875000
INFO  | ??:??:?? 5 Bandwidth set to 250.000000
INFO  | ??:??:?? 5 Power output set to 22
DEBUG | ??:??:?? 5 Current limit set to 140.000000
DEBUG | ??:??:?? 5 Current limit set result 0
DEBUG | ??:??:?? 5 Setting DIO2 as RF switch
DEBUG | ??:??:?? 5 SX126X_RXEN not defined, defaulting to RADIOLIB_NC
DEBUG | ??:??:?? 5 SX126X_TXEN not defined, defaulting to RADIOLIB_NC
DEBUG | ??:??:?? 5 Using MCU pin -1 as RXEN and pin -1 as TXEN to control RF switching
INFO  | ??:??:?? 5 Set RX gain to boosted mode; result: 0
INFO  | ??:??:?? 5 SX1262 Radio init succeeded, using SX1262 radio
INFO  | ??:??:?? 5 Not using WIFI
DEBUG | ??:??:?? 5 (bw=250, sf=11, cr=4/5) packet symLen=8 ms, payloadSize=237, time 2000 ms
DEBUG | ??:??:?? 5 LoRA bitrate = 118.500000 bytes / sec
INFO  | ??:??:?? 5 PowerFSM init, USB power=0
DEBUG | ??:??:?? 5 Enter state: BOOT
[  5592][D][esp32-hal-cpu.c:244] setCpuFrequencyMhz(): PLL: 480 / 6 = 80 Mhz, APB: 80000000 Hz
DEBUG | ??:??:?? 5 [Power] Battery: usbPower=1, isCharging=1, batMv=4221, batPct=100
DEBUG | ??:??:?? 5 [Screen] Screen: Started...
DEBUG | ??:??:?? 5 [GPS] Probing for GPS at 9600 
INFO  | ??:??:?? 7 [GPS] Found a UBlox Module using baudrate 9600
DEBUG | ??:??:?? 7 [GPS] Module Info : 
DEBUG | ??:??:?? 7 [GPS] Soft version: 1.00 (59842)
DEBUG | ??:??:?? 7 [GPS] Hard version: 00070000
DEBUG | ??:??:?? 7 [GPS] Extensions:2
DEBUG | ??:??:?? 7 [GPS]   PROTVER 14.00
DEBUG | ??:??:?? 7 [GPS]   GPS;SBAS;GLO;QZSS
DEBUG | ??:??:?? 7 [GPS] Protocol Version:14.00
DEBUG | ??:??:?? 7 [GPS] ProtVer=14
DEBUG | ??:??:?? 7 [GPS] Setting GPS+SBAS
INFO  | ??:??:?? 7 [GPS] GNSS configured for GPS+SBAS. Pause for 0.75s before sending next command.
INFO  | ??:??:?? 9 [GPS] GNSS module configuration saved!
DEBUG | ??:??:?? 9 [GPS] WANT GPS=0
INFO  | ??:??:?? 9 [GPS] Setting GPS power=0
DEBUG | ??:??:?? 9 [GPS] GPS powerdown using GPS_EN_ACTIVE
DEBUG | ??:??:?? 9 [cardKB] Using I2C Bus 1 (the second one)
INFO  | ??:??:?? 9 [RangeTestModule] Range Test Module - Disabled
DEBUG | ??:??:?? 9 [PowerFSM] Enter state: ON
INFO  | ??:??:?? 9 [PowerFSM] Initialise the NimBLE bluetooth module
INFO  | ??:??:?? 11 [Screen] Done with boot screen...
DEBUG | ??:??:?? 11 [Screen] showing standard frames
DEBUG | ??:??:?? 11 [Screen] Module wants a UI Frame
DEBUG | ??:??:?? 11 [Screen] Showing 1 module frames
DEBUG | ??:??:?? 11 [Screen] Total frame count: 104
DEBUG | ??:??:?? 11 [Screen] Added modules.  numframes: 1
DEBUG | ??:??:?? 11 [Screen] Finished building frames. numframes: 7
INFO  | ??:??:?? 15 [EnvironmentTelemetryModule] Environment Telemetry: Initializing
ERROR | ??:??:?? 15 [EnvironmentTelemetryModule] setConfig BSEC2 code: -34
INFO  | ??:??:?? 15 [EnvironmentTelemetryModule] BME680 state read from /prefs/bsec.dat.
WARN  | ??:??:?? 15 [EnvironmentTelemetryModule] updateSubscription BSEC2 code: 14
INFO  | ??:??:?? 15 [EnvironmentTelemetryModule] Init sensor: BME680 with the BSEC Library version 2.5.0.2
DEBUG | ??:??:?? 15 [EnvironmentTelemetryModule] BME680Sensor::runOnce: bme680.status 14
WARN  | ??:??:?? 15 [EnvironmentTelemetryModule] Could not connect to detected BME680 sensor.
 Removing from nodeTelemetrySensorsMap.
DEBUG | ??:??:?? 25 [Power] Battery: usbPower=1, isCharging=1, batMv=4231, batPct=100
INFO  | ??:??:?? 35 [NodeInfoModule] Sending our nodeinfo to mesh (wantReplies=1)`

**Comment by HarukiToreda** [comment]
nevermind, this is an unrelated issue. mentioned on in 3938

**Comment by HarukiToreda** [comment]
this works

**Comment by HarukiToreda** [comment]
> > @HarukiToreda @thebentern can someone please test that ? i am currently out of cardkb. but copilot said this should absolutely work, no questions there! :-)
> 
> It works but I think something may have broken on the BME680 Sensor. It's disabling after detected. [EnvironmentTelemetryModule]
> 
> DEBUG | ??:??:?? 11 [Screen] Finished building frames. numframes: 7 INFO | ??:??:?? 15 [EnvironmentTelemetryModule] Environment Telemetry: Initializing ERROR | ??:??:?? 15 [EnvironmentTelemetryModule] setConfig BSEC2 code: -34 INFO | ??:??:?? 15 [EnvironmentTelemetryModule] BME680 state read from /prefs/bsec.dat. WARN | ??:??:?? 15 [EnvironmentTelemetryModule] updateSubscription BSEC2 code: 14 INFO | ??:??:?? 15 [EnvironmentTelemetryModule] Init sensor: BME680 with the BSEC Library version 2.5.0.2 DEBUG | ??:??:?? 15 [EnvironmentTelemetryModule] BME680Sensor::runOnce: bme680.status 14 WARN | ??:??:?? 15 [EnvironmentTelemetryModule] Could not connect to detected BME680 sensor. Removing from nodeTelemetrySensorsMap. DEBUG | ??:??:?? 25 [Power] Battery: usbPower=1, isCharging=1, batMv=4231, batPct=100 INFO | ??:??:?? 35 [NodeInfoModule] Sending our nodeinfo to mesh (wantReplies=1)`


nevermind, this is an unrelated issue. mentioned on in 3938. The keyboard works fine

**Comment by HarukiToreda** [comment]
could someone else please test. It works for me for both Wireless Stick Lite and Heltec V3

**Review by caveman99 — CHANGES_REQUESTED** [review]
I rewrote parts of the scanner to just scan for ID's specified. Somehow you reverted my deletes of your code, and now it's doing both approaches at the same time. 

My idea was to scan during cardKB init again if the cardkb was net detected before, not do it 10 times at the end of main.cpp ...

**Review by caveman99 — APPROVED** [review]

**Comment by caveman99** [comment]
@HarukiToreda can you test this latest code? it's working for me on the m5stack with cardkb.

**Comment by HarukiToreda** [comment]
> @HarukiToreda can you test this latest code? it's working for me on the m5stack with cardkb.

Just tested on both Heltec V3 and Wireless tracker with cardkbs. It works perfectly.

**Pull request merged**